File: backend-service/app/schemas/chart.py
from datetime import date

from pydantic import BaseModel, ConfigDict, Field


class ChartPeriod(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    from_date: date | None = Field(default=None, alias="from")
    to_date: date | None = Field(default=None, alias="to")
    available_days: int


class TemperaturePoint(BaseModel):
    date: date
    average: float
    minimum: float
    maximum: float


class HumidityPoint(BaseModel):
    date: date
    average: float | None = None


class PrecipitationPoint(BaseModel):
    date: date
    total: float


class WindPoint(BaseModel):
    date: date
    average: float | None = None
    maximum: float


class ChartDataResponse(BaseModel):
    period: ChartPeriod
    temperature: list[TemperaturePoint]
    humidity: list[HumidityPoint]
    precipitation: list[PrecipitationPoint]
    wind: list[WindPoint]

File: ui-service/src/types/chart.ts
/** GET /api/weather/charts?from=&to= -- ordered ascending by date. */
export interface ChartPeriod {
  from: string | null;
  to: string | null;
  available_days: number;
}

export interface TemperaturePoint {
  date: string;
  average: number;
  minimum: number;
  maximum: number;
}

export interface HumidityPoint {
  date: string;
  average: number | null;
}

export interface PrecipitationPoint {
  date: string;
  total: number;
}

export interface WindPoint {
  date: string;
  average: number | null;
  maximum: number;
}

export interface ChartData {
  period: ChartPeriod;
  temperature: TemperaturePoint[];
  humidity: HumidityPoint[];
  precipitation: PrecipitationPoint[];
  wind: WindPoint[];
}
